---
title: "S4 classes, exercises"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{S4 classes, exercises}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(SummarizedExperiment)
data("airway_small", package = "BasicR")

#airway_small

```

1. Extract genes with exons overlapping the region 16:80000103-81066719
2. Calculate the mean counts for these rows
3. Exclude the last sample 


```{r}
gr1 <- GRanges(
    seqnames = "16",
    ranges = IRanges(80000103, 81066719),
    )
gr1 <- GRanges("16:80000103-81066719")
se_1 <- subsetByOverlaps(airway_small, gr1, ignore.strand=TRUE)

rowRanges(se_1)
rowMeans(assay(se_1))
rowMeans(assay(se_1[,-ncol(se_1)]))

```

4. Use the dds object from previously
5. Find the rows, where the gene expression is 0 for all the samples
6. Exclude these rows.

```{r}
library("pasilla")
pasCts <- system.file("extdata",
                      "pasilla_gene_counts.tsv",
                      package="pasilla", mustWork=TRUE)
pasAnno <- system.file("extdata",
                       "pasilla_sample_annotation.csv",
                       package="pasilla", mustWork=TRUE)
cts <- as.matrix(read.csv(pasCts,sep="\t",row.names="gene_id"))
coldata <- read.csv(pasAnno, row.names=1)
coldata <- coldata[,c("condition","type")]

rownames(coldata) <- sub("fb", "", rownames(coldata))
cts <- cts[, rownames(coldata)]
```

```{r}
library("DESeq2")
dds <- DESeqDataSetFromMatrix(countData = cts,
                              colData = coldata,
                              design = ~ condition)
dds
dds <- dds[1:200,]
dds <- DESeq(dds)
res <- results(dds)
res
```

```{r}
dds <- dds[-which(apply(assay(dds), 1, function(x) sum(x)==0)),]

```


7. Run the analysis again on this subset
```{r}

dds <- DESeq(dds)
```

8. Using the DeSeq2 vignette, find out how to do a PCA on the gene expression data. Tip: you will have to apply a variance stabilizing transformation on the count data.

```{r}
p <- prcomp(assay(dds))

plot(p$x)

vsd <- varianceStabilizingTransformation(dds, blind = FALSE)
plot(prcomp(assay(vsd))$x)

plotPCA(vsd)
```

